subctl/issue-89-reconstructing-broker-info.md

```diff
@@ -0,0 +1,70 @@
+# Recover broker-info.subm file from the running Broker
+
+Related Issue:
+[Option for recovering the broker-info.subm file from a running broker cluster](https://github.com/submariner-io/enhancements/issues/143)
+
+## Summary
+
+`broker-info.subm` file is created when a broker is installed. The file contains information critical to have a functional environment.
+There is a possibility that the file could be deleted or lost for one or the other reason. In such cases, no new clusters could be joined.
+
+This proposal describes a way to recover the `broker-info.subm` file from the running broker.
+
+## Proposal
+
+A new `subctl` command will be added that would fetch needed data from the broker and write it to `broker-info.subm` file. The file
+would be Base64 encoded.
+
+## Design details
+
+* subctl command name: `recover-brokerinfo`
+* Flags for the command:
+  * `submconfig` to specify kubeconfig of the cluster in which Submariner is installed
```

Comment:
As mentioned during the call, I'm personally not convinced that `submConfig` is an apt name as we use it in ACM/OCM. Please come up with a better name.

Reply:
That only works if we can recover everything using _only_ a `Submariner` resource; otherwise we need to be able to specify two contexts (one to access a broker, the other to access a joined cluster).

Reply:
The `Submariner` resource contains the connectivity info to access the broker, if needed.

Reply:
So, the original kubeconfig provided should be of Broker cluster. And if Submariner is not installed on that cluster, and there is a joined cluster, provide kubeconfig of the cluster that has Submariner installed. Right?

Reply:
As outlined in the steps I posted above, assuming we need the `Submariner` resource from a previously joined cluster in order to recover the broker info, then we expect the user to provide the kubeconfig of a joined cluster. If not and it's the broker cluster then we can try to recover iff there are no joined clusters.

Reply:
Client token is stored only on `broker`. So access to broker (ad hence its kubeconfig) is necessary. Unless we get broker restconfig from Submariner/service discovery like [gather does](https://github.com/submariner-io/subctl/blob/devel/internal/gather/gather.go#L183)

Reply:
Right - that's what I said above, also like the pods do.

Reply:
Turns out serviceaccounts are not accessible from the Submariner cluster.

Reply:
Back to the original comment, it seems we've deprecated the passing of kubeconfigs, so how about `--context` for the joined cluster and `--broker-context`.

Reply:
We haven’t deprecated the passing of kubeconfigs, we’ve made them consistent; so `--kubeconfig` and `--context` for the joined cluster, `--brokerconfig` and `--brokercontext` for the broker (following existing practice).